test/base.py

```diff
@@ -0,0 +1,46 @@
+import os
+import shutil
+import tempfile
+from pathlib import Path
+from unittest import TestCase
+
+
+class BaseTestCase(TestCase):
+    """Base class for our tests to house shared logic."""
+    _using_fake_hook_tools = False
+    _tmpdir = None
+
+    @property
+    def tmpdir(self):
+        if not self._tmpdir:
+            self._tmpdir = Path(tempfile.mkdtemp())
+            self.addCleanup(shutil.rmtree, self._tmpdir)
+        return self._tmpdir
+
+    def patch_env(self, **kwargs):
+        """Patch os.environ for a test, and ensure it gets restored.
+
+        Can be called multiple times.
+        """
+        old_env = dict(os.environ)
+        os.environ.update(kwargs)
+        self.addCleanup(os.environ.update, old_env)
+
+    def create_hook_tool(self, tool_name, output='', exit_code=0, impl=None):
```

Comment:
I appreciate that the intention here is to make it more convenient to use. The trade-off, though, is making the API more complex and error prone, and existing logic harder to extend. In other words, if we have `"echo foo\nexit 1"`, it's very clear what that does and how to go from here. But `create("foo", "echo foo")` is a mistake, and the fact the parameters of the function don't play well with each other, although they are all related, will also make things a bit more surprising.